bionic/flow.py

```diff
@@ -114,22 +114,25 @@ def add_case(self, name, case_key, value):
     def clear_providers(self, names):
         state = self
 
-        # Remember the original protocol for each provider.
-        protocols_by_entity_name = {}
+        # Remember the original protocol and docstring for each provider.
+        attrs_by_entity_name = {}
         for name in names:
             if not state.has_provider(name):
                 continue
             provider = self.get_provider(name)
-            for res_name, res_protocol in zip(
-                    provider.attrs.names, provider.attrs.protocols):
-                protocols_by_entity_name[res_name] = res_protocol
+
+            for res_name, res_protocol, res_docstring in zip(
+                    provider.attrs.names,
+                    provider.attrs.protocols,
+                    [provider.attrs.docstring]):
+                attrs_by_entity_name[res_name] = (res_protocol, res_docstring)
```

Comment:
Maybe it would be more straightforward to just store the entire `attrs` object.  That would make the name of this dict more precise, and would also make the code on line 135 clearer.

Reply:
Makes sense, we're updating!